# INFOSYS 722 Iteration 4 – BDAS
Diabetes prediction on **AWS EC2 + Apache Spark + PySpark + Spark MLlib**.
Pipeline: 25 features (19 base + 6 constructed), training-only scaling, stored 70:30 Partition, class-weighted LR/DT/RF, GBT (50 iter), KMeans k=5.
Run from the project root (`INFOSYS722-BDAS-Diabetes/`).


In [1]:
import sys
sys.path.insert(0, "src")
from pyspark.sql import SparkSession


def get_spark():

    spark = (
        SparkSession.builder
        .appName(
            "INFOSYS722-BDAS-Diabetes"
        )
        .getOrCreate()
    )

    return spark
spark = get_spark()
print("Spark Version:", spark.version)


Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


/home/ubuntu/.local/lib/python3.12/site-packages/pyspark/testing/utils.py:127: FutureWarning: PySpark does not yet fully support pandas >= 3.0.0. Some features may not work correctly. It is recommended to use pandas < 3.0.0 for now.
  require_minimum_pandas_version()
26/10/05 03:17:21 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Spark Version: 4.2.0


## Step 3 – Data Preparation
Load D1a, filter target-missing rows, construct BMI_band / health-day bands / RiskTest5 / flag / ComorbidityCount, assign the seeded Partition column, audit and persist `clean_diabetes.parquet` (432,339 × 28).


In [2]:
from spark_session import get_spark

from pyspark.sql.functions import (
    col,
    when,
    lit,
    rand
)


DATA_PATH = "data/brfss2023_diabetes_analysis.csv"


spark = get_spark()


print("==============================")
print("STEP 3 - DATA PREPARATION")
print("==============================")


# =====================================================
# 1. LOAD DATA
# =====================================================

df = (
    spark.read
    .option("header", True)
    .option("inferSchema", True)
    .csv(DATA_PATH)
)


print("\nOriginal dataset:")
print("Rows :", df.count())
print("Cols :", len(df.columns))


df.printSchema()



# =====================================================
# 2. REMOVE ONLY INVALID TARGET
# =====================================================
# Diabetes_binary adalah target.
# Record tanpa target tidak berguna untuk supervised learning.

df_clean = df.filter(
    col("Diabetes_binary").isNotNull()
)


print("\nAfter target cleaning:")
print("Rows :", df_clean.count())



# =====================================================
# 3. CREATE BMI BAND
# =====================================================

df_clean = df_clean.withColumn(
    "BMI_band",
    when(col("BMI") < 18.5, 1)
    .when(col("BMI") < 25, 2)
    .when(col("BMI") < 30, 3)
    .otherwise(4)
)



# =====================================================
# 4. CREATE HEALTH DAY BANDS
# =====================================================

df_clean = df_clean.withColumn(
    "MentHlth_band",
    when(col("MentHlth") == 0, 0)
    .when(col("MentHlth") <= 14, 1)
    .otherwise(2)
)


df_clean = df_clean.withColumn(
    "PhysHlth_band",
    when(col("PhysHlth") == 0, 0)
    .when(col("PhysHlth") <= 14, 1)
    .otherwise(2)
)



# =====================================================
# 5. CREATE RiskTest5
# =====================================================
#
# Approximation of CDC risk test:
# Age
# Sex
# BMI
# HighBP
# PhysicalActivity
#

df_clean = df_clean.withColumn(
    "RiskTest5",

    (
        when(col("Age") >= 9, 1)
        .otherwise(0)

        +

        when(col("Sex") == 0, 1)
        .otherwise(0)

        +

        when(col("BMI") >= 30, 1)
        .otherwise(0)

        +

        when(col("HighBP") == 1, 1)
        .otherwise(0)

        +

        when(col("PhysActivity") == 0, 1)
        .otherwise(0)
    )
)



df_clean = df_clean.withColumn(
    "RiskTest5_flag",
    when(col("RiskTest5") >= 5, 1)
    .otherwise(0)
)



# =====================================================
# 6. CREATE COMORBIDITY COUNT
# =====================================================

df_clean = df_clean.withColumn(
    "ComorbidityCount",

    (
        when(col("Stroke") == 1, 1)
        .otherwise(0)

        +

        when(col("HeartDiseaseorAttack") == 1, 1)
        .otherwise(0)

        +

        when(col("HighBP") == 1, 1)
        .otherwise(0)

        +

        when(col("HighChol") == 1, 1)
        .otherwise(0)
    )
)



# =====================================================
# 7. CREATE TRAINING / TESTING PARTITION
# =====================================================
#
# Keep partition before modelling.
# No randomSplit during ML stage.

df_clean = df_clean.withColumn(
    "Partition",
    when(
        rand(seed=42) <= 0.7,
        "Training"
    )
    .otherwise("Testing")
)



# =====================================================
# 8. DATA AUDIT
# =====================================================

print("\nFinal prepared dataset:")
print("Rows :", df_clean.count())
print("Cols :", len(df_clean.columns))


print("\nPartition:")
df_clean.groupBy(
    "Partition"
).count().show()


print("\nTarget distribution:")
df_clean.groupBy(
    "Diabetes_binary"
).count().show()



# =====================================================
# 9. SAVE OUTPUT
# =====================================================

(
    df_clean
    .write
    .mode("overwrite")
    .parquet(
        "data/clean_diabetes.parquet"
    )
)


print(
    "\nSaved: data/clean_diabetes.parquet"
)




STEP 3 - DATA PREPARATION



Original dataset:


Rows : 433323
Cols : 21
root
 |-- ID: integer (nullable = true)
 |-- Diabetes_binary: integer (nullable = true)
 |-- HighBP: integer (nullable = true)
 |-- HighChol: integer (nullable = true)
 |-- CholCheck: integer (nullable = true)
 |-- BMI: double (nullable = true)
 |-- Smoker: integer (nullable = true)
 |-- Stroke: integer (nullable = true)
 |-- HeartDiseaseorAttack: integer (nullable = true)
 |-- PhysActivity: integer (nullable = true)
 |-- HvyAlcoholConsump: integer (nullable = true)
 |-- AnyHealthcare: integer (nullable = true)
 |-- NoDocbcCost: integer (nullable = true)
 |-- GenHlth: integer (nullable = true)
 |-- MentHlth: integer (nullable = true)
 |-- PhysHlth: integer (nullable = true)
 |-- DiffWalk: integer (nullable = true)
 |-- Sex: integer (nullable = true)
 |-- Age: integer (nullable = true)
 |-- Education: integer (nullable = true)
 |-- Income: integer (nullable = true)


After target cleaning:


Rows : 432339



Final prepared dataset:


Rows : 432339
Cols : 28

Partition:


+---------+------+
|Partition| count|
+---------+------+
| Training|302683|
|  Testing|129656|
+---------+------+


Target distribution:


+---------------+------+
|Diabetes_binary| count|
+---------------+------+
|              1| 70380|
|              0|361959|
+---------------+------+



26/10/05 03:17:34 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.



Saved: data/clean_diabetes.parquet


## Step 4 – Data Transformation
Audit nulls over 25 features, fillna(0), VectorAssembler → raw_features, StandardScaler fitted on Training only → features; persist `model_ready.parquet`.


In [3]:
from spark_session import get_spark

from pyspark.sql.functions import col

from pyspark.ml.feature import (
    VectorAssembler,
    StandardScaler
)


spark = get_spark()


print("==============================")
print("STEP 4 - DATA TRANSFORMATION")
print("==============================")


# =====================================================
# LOAD PREPARED DATA
# =====================================================

df = spark.read.parquet(
    "data/clean_diabetes.parquet"
)


target = "Diabetes_binary"



# =====================================================
# REMOVE NON-FEATURE COLUMNS
# =====================================================

exclude_columns = [
    "ID",
    "Diabetes_binary",
    "Partition"
]


features = [
    c for c in df.columns
    if c not in exclude_columns
]


print("\nSelected features:")

for f in features:
    print(f)


print("\nNumber of features:")
print(len(features))



# =====================================================
# CHECK NULL VALUES
# =====================================================

print("\nNull value check:")

for c in features:

    null_count = (
        df
        .filter(
            col(c).isNull()
        )
        .count()
    )

    if null_count > 0:
        print(
            c,
            ":",
            null_count
        )



# =====================================================
# HANDLE REMAINING NULL VALUES
# =====================================================
#
# Missing values are handled after feature selection.
# This prevents leakage into unnecessary columns.
#

df = df.fillna(
    0,
    subset=features
)



# =====================================================
# VECTOR ASSEMBLY
# =====================================================

assembler = VectorAssembler(
    inputCols=features,
    outputCol="raw_features",
    handleInvalid="error"
)


df_vector = assembler.transform(
    df
)



print("\nVectorAssembler completed")



# =====================================================
# STANDARD SCALING
# =====================================================
#
# Fit scaler only using Training data.
# Testing data remains unseen.
#

training_data = (
    df_vector
    .filter(
        col("Partition") == "Training"
    )
)


scaler = StandardScaler(
    inputCol="raw_features",
    outputCol="features",
    withMean=True,
    withStd=True
)


scaler_model = scaler.fit(
    training_data
)



df_scaled = scaler_model.transform(
    df_vector
)



# =====================================================
# FINAL MODEL DATASET
# =====================================================

df_model = df_scaled.select(
    "features",
    "Diabetes_binary",
    "Partition",
    "Income",
    "RiskTest5",
    "RiskTest5_flag",
    "ComorbidityCount"
)



print("\nModel ready:")
df_model.show(5)



# =====================================================
# SAVE OUTPUT
# =====================================================

(
    df_model
    .write
    .mode("overwrite")
    .parquet(
        "data/model_ready.parquet"
    )
)


print(
    "\nSaved: data/model_ready.parquet"
)




STEP 4 - DATA TRANSFORMATION



Selected features:
HighBP
HighChol
CholCheck
BMI
Smoker
Stroke
HeartDiseaseorAttack
PhysActivity
HvyAlcoholConsump
AnyHealthcare
NoDocbcCost
GenHlth
MentHlth
PhysHlth
DiffWalk
Sex
Age
Education
Income
BMI_band
MentHlth_band
PhysHlth_band
RiskTest5
RiskTest5_flag
ComorbidityCount

Number of features:
25

Null value check:


HighBP : 1719
HighChol : 54660


CholCheck : 26219


BMI : 40220
Smoker : 22388


Stroke : 1220
HeartDiseaseorAttack : 4317


PhysActivity : 1196
HvyAlcoholConsump : 32600


AnyHealthcare : 18458
NoDocbcCost : 1469


GenHlth : 1222
MentHlth : 7984


PhysHlth : 10672
DiffWalk : 17673


Age : 7583
Education : 2200


Income : 86115



VectorAssembler completed



Model ready:


+--------------------+---------------+---------+------+---------+--------------+----------------+
|            features|Diabetes_binary|Partition|Income|RiskTest5|RiskTest5_flag|ComorbidityCount|
+--------------------+---------------+---------+------+---------+--------------+----------------+
|[1.20948539207589...|              1| Training|     0|        5|             1|               1|
|[1.20948539207589...|              0| Training|     0|        3|             0|               2|
|[1.20948539207589...|              0|  Testing|     2|        3|             0|               2|
|[-0.8267951831127...|              0| Training|     0|        2|             0|               0|
|[1.20948539207589...|              1| Training|     7|        3|             0|               1|
+--------------------+---------------+---------+------+---------+--------------+----------------+
only showing top 5 rows



Saved: data/model_ready.parquet


## Steps 5–7 – Method, Algorithms & Data Mining
**Binary classification** (LR maxIter=100 weighted, DT depth-5, RF 100 trees depth-8, GBT 50 iter) plus **KMeans k=5** on the 302,683-row Training partition; all five estimators scored on the 129,656-row test partition.


In [4]:
from spark_session import get_spark

from pyspark.sql.functions import (
    col,
    when
)

from pyspark.ml.classification import (
    LogisticRegression,
    DecisionTreeClassifier,
    RandomForestClassifier,
    GBTClassifier
)

from pyspark.ml.clustering import KMeans


spark = get_spark()


print("==============================")
print("STEP 7 - DATA MINING")
print("==============================")


# =====================================================
# LOAD MODEL READY DATA
# =====================================================

df = spark.read.parquet(
    "data/model_ready.parquet"
)


# =====================================================
# TRAIN / TEST SPLIT
# =====================================================

train = df.filter(
    col("Partition") == "Training"
)


test = df.filter(
    col("Partition") == "Testing"
)


print("\nTraining rows:")
print(train.count())


print("\nTesting rows:")
print(test.count())



# =====================================================
# CREATE CLASS WEIGHT
# =====================================================
#
# Handle imbalance:
# minority class receives higher weight
#

class_count = (
    train
    .groupBy("Diabetes_binary")
    .count()
    .collect()
)


counts = {}

for row in class_count:
    counts[row["Diabetes_binary"]] = row["count"]


majority = max(counts.values())
minority = min(counts.values())


minority_weight = majority / minority


print("\nClass counts:")
print(counts)


print("\nMinority class weight:")
print(minority_weight)



train_weighted = train.withColumn(
    "weight",
    when(
        col("Diabetes_binary") == 1,
        minority_weight
    )
    .otherwise(1.0)
)



# =====================================================
# 1. LOGISTIC REGRESSION
# =====================================================

lr = LogisticRegression(
    featuresCol="features",
    labelCol="Diabetes_binary",
    weightCol="weight",
    maxIter=100
)


lr_model = lr.fit(
    train_weighted
)


lr_result = lr_model.transform(
    test
)


print("\nLogistic completed")



# =====================================================
# 2. DECISION TREE
# =====================================================

dt = DecisionTreeClassifier(
    featuresCol="features",
    labelCol="Diabetes_binary",
    weightCol="weight",
    maxDepth=5,
    seed=42
)


dt_model = dt.fit(
    train_weighted
)


dt_result = dt_model.transform(
    test
)


print("Decision Tree completed")



# =====================================================
# 3. RANDOM FOREST
# =====================================================

rf = RandomForestClassifier(
    featuresCol="features",
    labelCol="Diabetes_binary",
    weightCol="weight",
    numTrees=100,
    maxDepth=8,
    seed=42
)


rf_model = rf.fit(
    train_weighted
)


rf_result = rf_model.transform(
    test
)


print("Random Forest completed")



# =====================================================
# 4. GRADIENT BOOSTED TREE
# =====================================================

gbt = GBTClassifier(
    featuresCol="features",
    labelCol="Diabetes_binary",
    maxIter=50,
    maxDepth=5,
    seed=42
)


gbt_model = gbt.fit(
    train_weighted
)


gbt_result = gbt_model.transform(
    test
)


print("GBT completed")



# =====================================================
# SAVE CLASSIFICATION OUTPUT
# =====================================================

lr_result.write \
    .mode("overwrite") \
    .parquet(
        "data/result_logistic.parquet"
    )


dt_result.write \
    .mode("overwrite") \
    .parquet(
        "data/result_tree.parquet"
    )


rf_result.write \
    .mode("overwrite") \
    .parquet(
        "data/result_rf.parquet"
    )


gbt_result.write \
    .mode("overwrite") \
    .parquet(
        "data/result_gbt.parquet"
    )



# =====================================================
# 5. KMEANS CLUSTERING
# =====================================================
#
# Unsupervised descriptive analysis
#

kmeans = KMeans(
    k=5,
    seed=42,
    featuresCol="features"
)


kmeans_model = kmeans.fit(
    train
)


cluster_result = kmeans_model.transform(
    train
)


cluster_result.write \
    .mode("overwrite") \
    .parquet(
        "data/result_cluster.parquet"
    )


print("KMeans completed")



# =====================================================
# FEATURE IMPORTANCE OUTPUT
# =====================================================


print("\nRandom Forest Feature Importance")

print(
    rf_model.featureImportances
)


print("\nDecision Tree Feature Importance")

print(
    dt_model.featureImportances
)



print("\nALL MODELS COMPLETED")




STEP 7 - DATA MINING

Training rows:


302683

Testing rows:
129656



Class counts:
{1: 49275, 0: 253408}

Minority class weight:
5.142729578893962



Logistic completed


Decision Tree completed


26/10/05 03:18:33 WARN DAGScheduler: Broadcasting large task binary with size 1868.0 KiB


26/10/05 03:18:44 WARN DAGScheduler: Broadcasting large task binary with size 3.6 MiB


26/10/05 03:18:57 WARN DAGScheduler: Broadcasting large task binary with size 1090.4 KiB


Random Forest completed


GBT completed


26/10/05 03:19:43 WARN DAGScheduler: Broadcasting large task binary with size 2.4 MiB


KMeans completed

Random Forest Feature Importance
(25,[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24],[0.12718418319069205,0.044121648170271865,0.01345910315991285,0.03938321704506947,0.000401468265682338,0.000583977342071292,0.008287084024668132,0.01059074812457514,0.00365746276336749,0.0003195628550492668,0.000240928544220074,0.15403382910477886,0.0014950900421214091,0.005700249942949287,0.05261383722511477,0.0068036859042239954,0.1164736481317477,0.0043403196625925195,0.007609959341427936,0.03353750828386335,0.0008372502696463943,0.008493609732759196,0.11509853753871487,0.0005388518292000586,0.24419423950527946])

Decision Tree Feature Importance
(25,[3,7,11,15,16,22,24],[0.016106726982985833,0.00283094026913113,0.21355361862101407,0.0033143625558512265,0.07922880261432035,0.01311613656884315,0.6718494123878541])

ALL MODELS COMPLETED


## Step 8 – Interpretation & Evaluation
Per-model AUC, accuracy, weighted precision/recall, F1, confusion matrices, top-20 gain/lift, four-model comparison CSV, and KMeans silhouette.


In [5]:
from spark_session import get_spark


from pyspark.sql.functions import (
    col,
    desc,
    udf
)


from pyspark.sql.types import DoubleType


from pyspark.ml.linalg import Vector


from pyspark.ml.evaluation import (
    BinaryClassificationEvaluator,
    MulticlassClassificationEvaluator,
    ClusteringEvaluator
)



spark = get_spark()



print("==============================")
print("STEP 8 - INTERPRETATION")
print("==============================")



# =====================================================
# EXTRACT POSITIVE PROBABILITY
# =====================================================

def extract_probability(v):

    if v is None:
        return 0.0

    return float(v[1])



probability_udf = udf(
    extract_probability,
    DoubleType()
)



# =====================================================
# MODEL FILES
# =====================================================


models = {

    "Logistic Regression":
        "data/result_logistic.parquet",

    "Decision Tree":
        "data/result_tree.parquet",

    "Random Forest":
        "data/result_rf.parquet",

    "GBT":
        "data/result_gbt.parquet"

}



auc_eval = BinaryClassificationEvaluator(
    labelCol="Diabetes_binary",
    rawPredictionCol="rawPrediction",
    metricName="areaUnderROC"
)



results = []



# =====================================================
# CLASSIFICATION EVALUATION
# =====================================================


for name, path in models.items():


    print("\n==============================")
    print(name)
    print("==============================")


    df = spark.read.parquet(path)



    auc = auc_eval.evaluate(df)



    accuracy = MulticlassClassificationEvaluator(
        labelCol="Diabetes_binary",
        predictionCol="prediction",
        metricName="accuracy"
    ).evaluate(df)



    precision = MulticlassClassificationEvaluator(
        labelCol="Diabetes_binary",
        predictionCol="prediction",
        metricName="weightedPrecision"
    ).evaluate(df)



    recall = MulticlassClassificationEvaluator(
        labelCol="Diabetes_binary",
        predictionCol="prediction",
        metricName="weightedRecall"
    ).evaluate(df)



    f1 = MulticlassClassificationEvaluator(
        labelCol="Diabetes_binary",
        predictionCol="prediction",
        metricName="f1"
    ).evaluate(df)



    print("AUC:", auc)
    print("Accuracy:", accuracy)
    print("Precision:", precision)
    print("Recall:", recall)
    print("F1:", f1)



    print("\nConfusion Matrix")

    df.groupBy(
        "Diabetes_binary",
        "prediction"
    ).count().show()



    # =================================================
    # TOP 20 GAIN / LIFT
    # =================================================


    ranked = (
        df
        .withColumn(
            "positive_probability",
            probability_udf(
                col("probability")
            )
        )
        .orderBy(
            desc("positive_probability")
        )
    )



    total_records = ranked.count()


    top20_records = int(
        total_records * 0.2
    )


    top20 = ranked.limit(
        top20_records
    )



    total_positive = (
        ranked
        .filter(
            col("Diabetes_binary") == 1
        )
        .count()
    )



    top20_positive = (
        top20
        .filter(
            col("Diabetes_binary") == 1
        )
        .count()
    )



    gain = (
        top20_positive /
        total_positive
    )


    lift = (
        gain /
        0.2
    )



    print(
        "Top 20% Gain:",
        gain
    )


    print(
        "Top 20% Lift:",
        lift
    )



    results.append(
        (
            name,
            float(auc),
            float(accuracy),
            float(precision),
            float(recall),
            float(f1),
            float(gain),
            float(lift)
        )
    )



# =====================================================
# SAVE MODEL COMPARISON
# =====================================================


metric_df = spark.createDataFrame(
    results,
    [
        "Model",
        "AUC",
        "Accuracy",
        "Precision",
        "Recall",
        "F1",
        "Top20_Gain",
        "Top20_Lift"
    ]
)



print("\nMODEL COMPARISON")

metric_df.show()



metric_df.write \
    .mode("overwrite") \
    .option(
        "header",
        True
    ) \
    .csv(
        "data/model_metrics"
    )



# =====================================================
# KMEANS
# =====================================================


print("\n==============================")
print("KMEANS EVALUATION")
print("==============================")


cluster_df = spark.read.parquet(
    "data/result_cluster.parquet"
)



cluster_df.groupBy(
    "prediction"
).count().show()



silhouette = ClusteringEvaluator(
    featuresCol="features",
    predictionCol="prediction",
    metricName="silhouette"
).evaluate(
    cluster_df
)



print(
    "Silhouette Score:",
    silhouette
)



print("\nEVALUATION COMPLETED")





/home/ubuntu/.local/lib/python3.12/site-packages/pyspark/sql/udf.py:116: FutureWarning: PySpark does not yet fully support pandas >= 3.0.0. Some features may not work correctly. It is recommended to use pandas < 3.0.0 for now.
  require_minimum_pandas_version()
/home/ubuntu/.local/lib/python3.12/site-packages/pyspark/sql/udf.py:120: RuntimeWarning: Arrow optimization failed to enable because PyArrow or Pandas is not installed. Falling back to a non-Arrow-optimized UDF.
  warnings.warn(


STEP 8 - INTERPRETATION

Logistic Regression


AUC: 0.8044790436489804
Accuracy: 0.718177330782995
Precision: 0.8385948770749493
Recall: 0.7181773307829951
F1: 0.7527267142652831

Confusion Matrix


+---------------+----------+-----+
|Diabetes_binary|prediction|count|
+---------------+----------+-----+
|              1|       0.0| 5260|
|              0|       0.0|77271|
|              1|       1.0|15845|
|              0|       1.0|31280|
+---------------+----------+-----+



Top 20% Gain: 0.5138118929163705
Top 20% Lift: 2.5690594645818523

Decision Tree


AUC: 0.7055394415215654
Accuracy: 0.688259702597643
Precision: 0.8373927739121847
Recall: 0.688259702597643
F1: 0.7282873462423894

Confusion Matrix
+---------------+----------+-----+
|Diabetes_binary|prediction|count|
+---------------+----------+-----+
|              1|       0.0| 4729|
|              0|       0.0|72861|
|              1|       1.0|16376|
|              0|       1.0|35690|
+---------------+----------+-----+



Top 20% Gain: 0.47633262260127934
Top 20% Lift: 2.3816631130063968

Random Forest


AUC: 0.8037650935241546
Accuracy: 0.7001758499413834
Precision: 0.8410643801952868
Recall: 0.7001758499413833
F1: 0.7383119899527009

Confusion Matrix
+---------------+----------+-----+
|Diabetes_binary|prediction|count|
+---------------+----------+-----+
|              1|       0.0| 4623|
|              0|       0.0|74300|
|              1|       1.0|16482|
|              0|       1.0|34251|
+---------------+----------+-----+



Top 20% Gain: 0.512106135986733
Top 20% Lift: 2.560530679933665

GBT


AUC: 0.811132901455354
Accuracy: 0.8437249336706362
Precision: 0.8109678244443471
Recall: 0.8437249336706361
F1: 0.8086910414160577

Confusion Matrix
+---------------+----------+------+
|Diabetes_binary|prediction| count|
+---------------+----------+------+
|              1|       0.0| 17259|
|              0|       0.0|105548|
|              1|       1.0|  3846|
|              0|       1.0|  3003|
+---------------+----------+------+



Top 20% Gain: 0.52030324567638
Top 20% Lift: 2.6015162283818998

MODEL COMPARISON


+-------------------+------------------+------------------+------------------+------------------+------------------+-------------------+------------------+
|              Model|               AUC|          Accuracy|         Precision|            Recall|                F1|         Top20_Gain|        Top20_Lift|
+-------------------+------------------+------------------+------------------+------------------+------------------+-------------------+------------------+
|Logistic Regression|0.8044790436489804| 0.718177330782995|0.8385948770749493|0.7181773307829951|0.7527267142652831| 0.5138118929163705|2.5690594645818523|
|      Decision Tree|0.7055394415215654| 0.688259702597643|0.8373927739121847| 0.688259702597643|0.7282873462423894|0.47633262260127934|2.3816631130063968|
|      Random Forest|0.8037650935241546|0.7001758499413834|0.8410643801952868|0.7001758499413833|0.7383119899527009|  0.512106135986733| 2.560530679933665|
|                GBT| 0.811132901455354|0.8437249336706362|0.810


KMEANS EVALUATION


+----------+------+
|prediction| count|
+----------+------+
|         1| 88024|
|         3| 49919|
|         4| 40129|
|         2| 12759|
|         0|111852|
+----------+------+



Silhouette Score: 0.15494829834321938

EVALUATION COMPLETED


In [6]:
spark.stop()
